File: src/pages/GRCPage.java
// ============================================
// 🤖 AI SELF-HEALING FIX
// ============================================
// Timestamp: 2026-02-19T07:58:19.524Z
// File: src/pages/GRCPage.java
// Field: emailPhoneInput
//
// SELECTOR FIXED:
// Old: #login-idd
// New: #login-id
// Exception reported: #login-idd
//
// Confidence: 85%
// Replacements: 1
// ============================================

package src.pages;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.WaitForSelectorState;

/**
 * Page Object Model class for GRC Login page
 * Handles login functionality with phone/email, OTP generation, and password login
 */
public class GRCPage {
    private final Page page;
    
    // Locators
    private final String emailPhoneInput = "#login-id";
    private final String getOtpButton = "button:has-text('Get OTP')";
    private final String loginWithPasswordLink = "button:has-text('Login with Password')";
    private final String signUpLink = "a[href='/grc/auth/signup']";
    private final String errorMessage = ".text-red-500";
    private final String emailLabel = "label[for='login-id']";
    private final String loginForm = "form";
    
    public GRCPage(Page page) {
        this.page = page;
    }
    
    /**
     * Navigate to GRC login page
     * @param url The URL to navigate to
     */
    public void navigateTo(String url) {
        page.navigate(url);
        page.waitForSelector(emailPhoneInput, new Page.WaitForSelectorOptions().setState(WaitForSelectorState.VISIBLE));
    }
    
    /**
     * Enter email or phone number in the login field
     * @param emailOrPhone Email address or phone number to enter
     */
    public void enterEmailOrPhone(String emailOrPhone) {
        page.fill(emailPhoneInput, emailOrPhone);
    }
    
    /**
     * Click the Get OTP button
     */
    public void clickGetOtpButton() {
        page.click(getOtpButton);
    }
    
    /**
     * Click the Login with Password link
     */
    public void clickLoginWithPasswordLink() {
        page.click(loginWithPasswordLink);
    }
    
    /**
     * Click the Sign Up link
     */
    public void clickSignUpLink() {
        page.click(signUpLink);
    }
    
    /**
     * Get the error message text when validation fails
     * @return Error message text
     */
    public String getErrorMessage() {
        return page.textContent(errorMessage);
    }
    
    /**
     * Check if error message is visible
     * @return true if error message is visible
     */
    public boolean isErrorMessageVisible() {
        return page.isVisible(errorMessage);
    }
    
    /**
     * Get the current value in the email/phone input field
     * @return Current input value
     */
    public String getEmailPhoneValue() {
        return page.inputValue(emailPhoneInput);
    }
    
    /**
     * Check if Get OTP button is enabled
     * @return true if button is enabled
     */
    public boolean isGetOtpButtonEnabled() {
        return page.isEnabled(getOtpButton);
    }
    
    /**
     * Check if Login with Password link is visible
     * @return true if link is visible
     */
    public boolean isLoginWithPasswordLinkVisible() {
        return page.isVisible(loginWithPasswordLink);
    }
    
    /**
     * Wait for page to load completely
     */
    public void waitForPageLoad() {
        page.waitForSelector(emailPhoneInput);
        page.waitForSelector(getOtpButton);
        page.waitForSelector(loginWithPasswordLink);
    }
    
    /**
     * Clear the email/phone input field
     */
    public void clearEmailPhoneField() {
        page.fill(emailPhoneInput, "");
    }
    
    /**
     * Get page title
     * @return Page title
     */
    public String getPageTitle() {
        return page.title();
    }
    
    /**
     * Check if the form is visible
     * @return true if login form is visible
     */
    public boolean isLoginFormVisible() {
        return page.isVisible(loginForm);
    }
}

File: src/pages/GRCPage.js
import { expect } from '@playwright/test';

/**
 * Page Object Model for GRC Login Page
 * Handles user authentication and login functionality
 */
export class GRCPage {
  constructor(page) {
    this.page = page;
    
    // Locators
    this.emailInput = page.locator('#login-id');
    this.emailLabel = page.locator('label[for="login-id"]');
    this.getOTPButton = page.getByRole('button', { name: 'Get OTP' });
    this.loginForm = page.locator('form');
    this.signUpLink = page.getByRole('link', { name: 'Sign Up' });
    this.loginWithPasswordButton = page.getByRole('button', { name: 'Login with Password' });
    this.errorMessage = page.locator('.text-red-500.text-\[1\.2rem\]');
    this.pageTitle = page.locator('h1');
    
    // Navigation elements
    this.zolvitLogo = page.locator('img[alt="zolvitLogo"]');
    this.termsOfServiceLink = page.getByRole('link', { name: 'Terms of service' });
    this.privacyPolicyLink = page.getByRole('link', { name: 'Privacy policy' });
  }

  /**
   * Navigate to the GRC login page
   */
  async navigate() {
    await this.page.goto('/grc/auth/signin');
    await this.waitForPageLoad();
  }

  /**
   * Wait for the page to be fully loaded
   */
  async waitForPageLoad() {
    await expect(this.pageTitle).toContainText('Log into your account');
    await expect(this.emailInput).toBeVisible();
    await expect(this.getOTPButton).toBeVisible();
  }

  /**
   * Enter email address or mobile number in the login field
   * @param {string} emailOrMobile - Email address or mobile number to enter
   */
  async enterEmailAddress(emailOrMobile) {
    await this.emailInput.waitFor({ state: 'visible' });
    await this.emailInput.clear();
    await this.emailInput.fill(emailOrMobile);
    
    // Verify the input was entered correctly
    await expect(this.emailInput).toHaveValue(emailOrMobile);
  }

  /**
   * Click the Get OTP button
   */
  async clickGetOTP() {
    await this.getOTPButton.waitFor({ state: 'visible' });
    await expect(this.getOTPButton).toBeEnabled();
    await this.getOTPButton.click();
  }

  /**
   * Complete the login flow by entering email and clicking Get OTP
   * @param {string} emailOrMobile - Email address or mobile number
   */
  async loginWithOTP(emailOrMobile) {
    await this.enterEmailAddress(emailOrMobile);
    await this.clickGetOTP();
  }

  /**
   * Click the Login with Password button
   */
  async clickLoginWithPassword() {
    await this.loginWithPasswordButton.waitFor({ state: 'visible' });
    await this.loginWithPasswordButton.click();
  }

  /**
   * Click the Sign Up link
   */
  async clickSignUp() {
    await this.signUpLink.waitFor({ state: 'visible' });
    await this.signUpLink.click();
  }

  /**
   * Get the current error message text
   * @returns {Promise<string>} Error message text
   */
  async getErrorMessage() {
    await this.errorMessage.waitFor({ state: 'visible' });
    return await this.errorMessage.textContent();
  }

  /**
   * Check if error message is displayed
   * @returns {Promise<boolean>} True if error message is visible
   */
  async isErrorMessageDisplayed() {
    try {
      await this.errorMessage.waitFor({ state: 'visible', timeout: 3000 });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Verify the email input has the correct styling (error state)
   */
  async verifyEmailInputErrorState() {
    await expect(this.emailInput).toHaveClass(/border-red-500/);
    await expect(this.emailInput).toHaveClass(/text-red-500/);
  }

  /**
   * Verify the email input is in normal state
   */
  async verifyEmailInputNormalState() {
    await expect(this.emailInput).not.toHaveClass(/border-red-500/);
    await expect(this.emailInput).not.toHaveClass(/text-red-500/);
  }

  /**
   * Get the placeholder text of the email input
   * @returns {Promise<string>} Placeholder text
   */
  async getEmailInputPlaceholder() {
    return await this.emailInput.getAttribute('placeholder');
  }

  /**
   * Verify page title and main elements are present
   */
  async verifyPageElements() {
    await expect(this.pageTitle).toContainText('Log into your account');
    await expect(this.emailInput).toBeVisible();
    await expect(this.getOTPButton).toBeVisible();
    await expect(this.loginWithPasswordButton).toBeVisible();
    await expect(this.signUpLink).toBeVisible();
  }

  /**
   * Wait for form submission to complete
   */
  async waitForFormSubmission() {
    // Wait for any loading states or navigation after form submission
    await this.page.waitForLoadState('networkidle');
  }
}